rmm/monitor/src/rmi/rtt.rs

```diff
@@ -51,19 +52,44 @@ pub fn set_event_handler(mainloop: &mut Mainloop) {
                 return Err(Error::RmiErrorRtt);
             }
         }
-        // TODO: update mapping
-        super::dummy();
+
+        if rec.ripas_state() != ripas as u8 {
+            return Err(Error::RmiErrorInput);
+        }
+
+        if rec.ripas_addr() != ipa as u64 {
+            return Err(Error::RmiErrorInput);
+        }
+
         let map_size = level_to_size(level);
+        if ipa as u64 + map_size > rec.ripas_end() {
+            return Err(Error::RmiErrorInput);
+        }
+
+        if ripas as u64 == RIPAS_EMPTY {
+            let ret = rmi.make_shared(rd.id(), ipa, level);
+            if ret.is_err() {
+                return Err(Error::RmiErrorInput);
+            }
+        } else if ripas as u64 == RIPAS_RAM {
+        }
         rec.inc_ripas_addr(map_size);
         Ok(())
     });
 
-    listen!(mainloop, rmi::RTT_READ_ENTRY, |arg, ret, _| {
-        super::dummy();
+    listen!(mainloop, rmi::RTT_READ_ENTRY, |arg, ret, rmm| {
+        let rmi = rmm.rmi;
+        let rd_granule = get_granule_if!(arg[0], GranuleState::RD)?;
+        let rd = rd_granule.content::<Rd>();
+        let ipa = arg[1];
+        let level = arg[2];
+
+        let res = rmi.rtt_read_entry(rd.id(), ipa, level)?;
+        ret[1] = res[0];
+        ret[2] = res[1];
+        ret[3] = res[2];
+        ret[4] = res[3];
```

Comment:
[minor] How about this?

```suggestion
        ret[1..5].clone_from_slice(&res[0..4]);
```

Reply:
Good suggestion. I've changed it with `copy_from_slice()`.